heareval/tasks/speech_commands.py

```diff
@@ -82,50 +184,182 @@ def run(self):
+
+        # Save the process metadata
+        process_metadata.to_csv(
+            os.path.join(self.workdir, self.outfile),
+            columns=PROCESSMETADATACOLS,
+            header=False,
+            index=False,
+        )
+
+        self.mark_complete()
+
+
+class ConfigureTestMetaData(WorkTask):
+    """
+    This config is data dependent and has to be set for each data
+    """
+
+    outfile = luigi.Parameter()
+
+    def requires(self):
+        return {
+            "test": ExtractArchiveTest(infile="test-corpus.tar.gz"),
+        }
+
+    @staticmethod
+    def name_for_slug(path):
+        # In the test set all the unknown sounds already have the class label
+        # prepended to the name, so we don't want to add 'unknown' to them as well.
+        label = os.path.basename(os.path.dirname(path))
+        if label == UNKNOWN:
+            path = os.path.basename(path)
+        return path
+
+    def run(self):
+        # The testing set is all the files in the separate testing download
+        test_path = Path(self.requires()["test"].workdir)
+        test_paths = [str(p.relative_to(test_path)) for p in test_path.glob("*/*.wav")]
+
+        process_metadata = pd.DataFrame(test_paths, columns=["relpath"])
+        process_metadata["partition"] = "test"
+
+        # Create a unique slug for each file. We include the folder with the class
+        # name b/c the base filenames may not be unique.
+        process_metadata["slug"] = (
+            process_metadata["relpath"]
+            .apply(self.name_for_slug)
+            .apply(slugify_file_name)
+        )
+        process_metadata["relpath"] = process_metadata["relpath"].apply(
+            partial(os.path.join, test_path)
+        )
+
+        # Hash the field id rather than the full path.
+        process_metadata["filename_hash"] = (
+            process_metadata["slug"]
+            .apply(lambda relpath: re.sub(r"-nohash-.*$", "", relpath))
+            .apply(filename_to_int_hash)
         )
 
         # Get label for the data from anywhere.
         # In this case it is the folder name
         process_metadata["label"] = process_metadata["relpath"].apply(
-            lambda relative_path: os.path.basename(os.path.dirname(relative_path))
+            ConfigureTrainValMetaData.apply_label
         )
 
         # Save the process metadata
         process_metadata.to_csv(
-            os.path.join(self.workdir, "process_metadata.csv"),
+            os.path.join(self.workdir, self.outfile),
             columns=PROCESSMETADATACOLS,
             header=False,
             index=False,
         )
 
-        with self.output().open("w") as _:
-            pass
+        self.mark_complete()
 
 
-class SubsampleCorpus(SubsampleCorpus):
+class CombineMetaData(WorkTask):
+    """
+    Because we dealt with the metadata for the train/val and testing datasets
+    separately this combines them together into a single csv file.
+    """
```

Comment:
`CombineMetaData` depends on the `ConfigureTestMetadata` and `ConfigureTrainValMetadata`. Mostly what the two-class for `test` and `trainval` are doing is to fetch the data for the splits followed by some operations which are independent of split(getting the `slug`, `label` and `file hash`). We can switch to the `ConfigureProcessMetadata` with the appropriate method to load these datasets and do the split independent tasks. I have worked on this class to this from one place. Will commit.

Reply:
Nice -- your refactors are excellent.